src/server/game/AI/PlayerAI/PlayerAI.cpp

```diff
@@ -94,12 +562,97 @@ bool PlayerAI::IsPlayerRangedAttacker(Player const* who)
             return false;
         }
         case CLASS_PRIEST:
-            return who->HasSpell(SPELL_SHADOWFORM);
+            return (PlayerAI::GetPlayerSpec(who) == SPEC_PRIEST_SHADOW);
         case CLASS_SHAMAN:
-            return !who->HasSpell(SPELL_STORMSTRIKE);
+            return (PlayerAI::GetPlayerSpec(who) == SPEC_SHAMAN_ELEMENTAL);
         case CLASS_DRUID:
-            return who->HasSpell(SPELL_MOONKIN_FORM);
+            return (PlayerAI::GetPlayerSpec(who) == SPEC_DRUID_BALANCE);
+    }
+}
+
+PlayerAI::TargetedSpell PlayerAI::VerifySpellCast(uint32 spellId, Unit* target)
+{
+    // Find highest spell rank that we know
+    uint32 knownRank, nextRank;
+    if (me->HasSpell(spellId))
+        // this will save us some lookups if the player has the highest rank (expected case)
+        nextRank = sSpellMgr->GetNextSpellInChain(knownRank = spellId);
+    else
+        knownRank = 0, nextRank = sSpellMgr->GetFirstSpellInChain(spellId);
+
+    while (nextRank && me->HasSpell(nextRank))
+        nextRank = sSpellMgr->GetNextSpellInChain(knownRank = nextRank);
+
+    if (!knownRank)
+        return {};
+
+    SpellInfo const* spellInfo = sSpellMgr->GetSpellInfo(knownRank);
+    if (!spellInfo)
+        return {};
+
+    if (me->GetSpellHistory()->HasGlobalCooldown(spellInfo))
+        return {};
+
+    Spell* spell = new Spell(me, spellInfo, TRIGGERED_NONE);
```

Comment:
Spell selection working on creating new Spell objects only to discard all but one is horribly bad. You need to redesign it so only the spell that will be cast gets constructed